crates/resolver/src/name_server/name_server.rs

```diff
@@ -35,74 +36,72 @@ pub struct NameServer<P: ConnectionProvider> {
-        config: ConnectionConfig,
+        config: NameServerConfig,
         options: Arc<ResolverOpts>,
         tls: Arc<TlsConfig>,
         connection_provider: P,
     ) -> Self {
-        Self {
-            inner: Arc::new(NameServerState::new(
-                server_config,
-                config,
-                options,
-                tls,
-                None,
-                connection_provider,
-            )),
-        }
+        Self::with_connections([], config, options, tls, connection_provider)
     }
 
     #[doc(hidden)]
-    pub fn from_conn(
-        server_config: &NameServerConfig,
-        config: ConnectionConfig,
+    pub fn with_connections(
+        connections: impl IntoIterator<Item = (Protocol, P::Conn)>,
+        config: NameServerConfig,
         options: Arc<ResolverOpts>,
         tls: Arc<TlsConfig>,
-        client: P::Conn,
         connection_provider: P,
     ) -> Self {
         Self {
             inner: Arc::new(NameServerState::new(
-                server_config,
+                connections
+                    .into_iter()
+                    .map(|(protocol, handle)| ConnectionState {
+                        protocol,
+                        handle,
+                        meta: Arc::new(ConnectionMeta::default()),
+                    })
+                    .collect(),
                 config,
                 options,
                 tls,
-                Some(client),
                 connection_provider,
             )),
         }
     }
 
     // TODO: there needs to be some way of customizing the connection based on EDNS options from the server side...
-    pub(super) async fn send(&self, request: DnsRequest) -> Result<DnsResponse, ProtoError> {
+    pub(super) fn send(
+        &self,
+        request: DnsRequest,
+        skip_udp: bool,
+    ) -> impl Future<Output = Result<DnsResponse, ProtoError>> {
         let this = self.clone();
         // if state is failed, return future::err(), unless retry delay expired..
-        this.inner.send(request).await
+        this.inner.send(request, skip_udp)
+    }
+
+    pub(super) fn decayed_srtt(&self) -> f64 {
+        self.inner.server_srtt.current()
     }
 
     #[cfg(test)]
     #[allow(dead_code)]
     pub(crate) fn is_connected(&self) -> bool {
-        use Status::*;
-        match (self.inner.status(), self.inner.client.try_lock()) {
-            (Established | Init, Some(client)) => client.is_some(),
-            (Failed, _) => false,
+        let Some(connections) = self.inner.connections.try_lock() else {
             // assuming that if someone has it locked it will be or is connected
-            (_, None) => true,
-        }
-    }
-
-    pub(super) fn decayed_srtt(&self) -> f64 {
-        self.inner.meta.srtt.current()
-    }
+            return true;
+        };
 
-    pub(super) fn protocol(&self) -> Protocol {
-        self.inner.config.protocol.to_protocol()
+        use Status::*;
```

Comment:
pre-existing, but this wildcard import doesn't feel worth its weight given the short prefix and the fact that only 3 variants are ref'd below.